[quant][fx] Validate dictionary keys

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58566 [quant][fx] Validate dictionary keys**

Summary: Validates the keys of the qconfig_dict, prepare_custom_config_dict, convert_custom_config_dict, and
fuse_custom_config_dict. If the user passes in an invalid key or makes a type, we will throw and error and let the user know what keys are supported.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #57509

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

angelayi committed:
[quant][fx] Validate qconfig_dict keys

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

angelayi committed:
Update on "[quant][fx] Validate qconfig_dict keys"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx] Validate qconfig_dict keys"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx] Validate qconfig_dict keys"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary: Validates the keys of the qconfig_dict, prepare_custom_config_dict, convert_custom_config_dict, and
fuse_custom_config_dict. If the user passes in an invalid key or makes a type, we will throw and error and let the user know what keys are supported.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #57509

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

vkuzo commented on torch/quantization/quantize_fx.py:
```diff
@@ -17,6 +17,97 @@ def _check_is_graph_module(model: torch.nn.Module) -> None:
             'sure to follow the tutorials.')
 
+
+def _check_is_valid_qconfig_dict(qconfig_dict: Any) -> None:
```
can we move these functions here (https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/qconfig_utils.py) instead, to keep LOC of quantize_fx.py manageable?

angelayi replied:
Sure! Should I also move _check_is_graph_module (lines 13-18) to qconfig_utils.py?

vkuzo replied:
great!

> Sure! Should I also move _check_is_graph_module (lines 13-18) to qconfig_utils.py?

sure, moving that to a utils file makes sense as well, with a couple of caveats:
* would be great to have that be a separate PR (since it's not technically related to this PR).  Also, this is optional so you are welcome to do this if you'd like but it's not required.
* since that util is not related to qconfig, maybe `torch/quantization/fx/utils.py` would be a good place for it

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary: Validates the keys of the qconfig_dict, prepare_custom_config_dict, convert_custom_config_dict, and
fuse_custom_config_dict. If the user passes in an invalid key or makes a type, we will throw and error and let the user know what keys are supported.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #57509

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

vkuzo commented:
looks great! Will accept after CI passes.  Could you reexport the new version to FB as well to run FB specific CI.

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary: Validates the keys of the qconfig_dict, prepare_custom_config_dict, convert_custom_config_dict, and
fuse_custom_config_dict. If the user passes in an invalid key or makes a type, we will throw and error and let the user know what keys are supported.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #57509

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

jerryzh168 commented on torch/quantization/fx/qconfig_utils.py:
```diff
@@ -97,3 +97,89 @@ def get_qconfig(qconfig_dict, module_type, module_name, global_qconfig):
+      `qconfig_dict`: dictionary whose keys we want to check
+    """
+
+    qconfig_dict_allowed_keys = {"", "object_type", "module_name_regex", "module_name"}
```
can we make this an argument of the function and make this a more general function instead of having 4 functions?

for example: we can have a `check_is_valid_config_dict(config_dict: Dict[str, Any], allowed_keys: Set[str])`

angelayi replied:
Ooh that's a good idea -- I will do so right now!

angelayi replied:
Do you think it would be better to still keep those 4 functions to pass in the specific allowed keys? This way, the specific allowed keys could be hidden from people looking into the quantize.py file.

Another idea could be to have the allowed keys as global variables in quantize.py which can then be referenced in the test cases. Right now I have just declared the list of allowed keys again in the test files. What are your thoughts on this?

jerryzh168 replied:
yeah, keeping the functions sounds OK I think

jerryzh168 replied:
I see, I think we can keep those 4 functions, also we can remove references to the list of allowed keys in test cases as well so that when we add new keys we don't need to update the test case.
I feel it's better to avoid having global variables in this case.

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary: Validates the keys of the qconfig_dict, prepare_custom_config_dict, convert_custom_config_dict, and
fuse_custom_config_dict. If the user passes in an invalid key or makes a type, we will throw and error and let the user know what keys are supported.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #57509

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

supriyar commented:
@angelayi  there seem to be some merge conflicts. Could you rebase to latest and fix them? Thanks!

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary: Validates the keys of the qconfig_dict, prepare_custom_config_dict, convert_custom_config_dict, and
fuse_custom_config_dict. If the user passes in an invalid key or makes a type, we will throw and error and let the user know what keys are supported.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #57509

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary: Validates the keys of the qconfig_dict, prepare_custom_config_dict, convert_custom_config_dict, and
fuse_custom_config_dict. If the user passes in an invalid key or makes a type, we will throw and error and let the user know what keys are supported.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #57509

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary: Validates the keys of the qconfig_dict, prepare_custom_config_dict, convert_custom_config_dict, and
fuse_custom_config_dict. If the user passes in an invalid key or makes a type, we will throw and error and let the user know what keys are supported.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #57509

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).

jerryzh168 reviewed (approved):
great progress! in tests, please add check for typo strings in the message as well

angelayi committed:
Update on "[quant][fx] Validate dictionary keys"


Summary: Validates the keys of the qconfig_dict, prepare_custom_config_dict, convert_custom_config_dict, and
fuse_custom_config_dict. If the user passes in an invalid key or makes a type, we will throw and error and let the user know what keys are supported.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #57509

Differential Revision: [D28540923](https://our.internmc.facebook.com/intern/diff/D28540923)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28540923).
